include/seastar/net/tls.hh

```diff
@@ -152,6 +152,10 @@ namespace tls {
      */
     using dn_callback = noncopyable_function<void(session_type type, sstring subject, sstring issuer)>;
 
+    class certificate_credentials;
+    void* get_impl(certificate_credentials const&);
```

Comment:
do i understand correctly that `get_impl` is only used in tests?

if that's true, im not sure what seastar devs will say, but at a minimum it'd probably worth stating that in the commit message, and having a comment in the test explaining what get_impl being null/non-null is testing.

Reply:
> do i understand correctly that `get_impl` is only used in tests?

Some tests should probably be added.

It is used to recover the credentials object in: https://github.com/redpanda-data/redpanda/pull/13477/files#diff-1b2288fb7841d76732d7e4973e09f8e8179bb09f0481364c71fc549355c74106R357-R358

Reply:
>Some tests should probably be added.

Separate test for `get_impl` then? It's used inline in the unit test I added for the new callback, but that's sort of informal.

What sort of test would be appropriate here? Something consuming the underlying `gnutls` APIs, or more like "returns `nullptr` (or not) when I expect it to"?

Reply:
>   auto g_creds = reinterpret_cast<gnutls_certificate_credentials_t>(get_impl(creds));

eeek! well i don't have a better suggestion.

Reply:
@oleiman now that i understand better, in the upstream PR i could just add the necessary context: we want to get access to all the gory bits of the certs and the "easiest" way to do that is to get access to the gnutls objects. definitely not sure what they'll suggest, but it seems reasonable to want this information in some capacity.

Reply:
> >Some tests should probably be added.
> 
> Separate test for `get_impl` then? It's used inline in the unit test I added for the new callback, but that's sort of informal.

Oh, I meant for the new callback. If there are tests for that, great.

I didn't get round to asking on the mailing list ideas for accessing the impl, or details about the underlying certs.

Reply:
I *think* that Evgeny proposed a PR upstream when we made changes to get the distinguished name that exposed the impl and they pushed back. I don't know if I'm remembering that correctly, but you could find it and see what they had to say (it's either a PR or on the mailing list).

The DN was easy because it was just a string. But now we want lots of other stuff.

Reply:
Sounds good. Merging this as is, then we can address any feedback from seastar guys as it comes up.